### PR TITLE
feat(windows): Windows boilerplate and CI (#2715)

### PR DESCRIPTION
**Pull request by ReactorScram**
Testing if CI will build the Windows exe, or at least check the code.

### CODE REVIEW

**Comment by ReactorScram** [comment]
This page says "rust / test" hasn't completed, but on the other page it says all checks passed?

File: .github/workflows/cd.yml
Comment (jamilbk) [inline .github/workflows/cd.yml:264]: This seems to be a matrix of dimension 1x1 -- could this be simplified to env vars or explicitly used in the remaining steps below? Or were you thinking this matrix would be expanded to cover more variables?

File: .github/workflows/cd.yml
Comment (jamilbk) [inline .github/workflows/cd.yml]: (nit) naming
```suggestion
  build-push-windows-release-artifacts:
```

File: .github/workflows/cd.yml
Comment (jamilbk) [inline .github/workflows/cd.yml:1]: Could you also rename `build-push-release-artifacts` to `build-push-linux-release-artifacts` in this file? You'll need to update it [here](https://github.com/firezone/firezone/pull/2730/files#diff-ea3ea8c9932adc7ba8161ceda844fedd43b006848ef1140c050cbd7ea0788a18R116) and [here](https://github.com/firezone/firezone/pull/2730/files#diff-ea3ea8c9932adc7ba8161ceda844fedd43b006848ef1140c050cbd7ea0788a18R305)

**Review by jamilbk — CHANGES_REQUESTED** [review]
Great start! Just some minor questions / suggestions around the CI config.

I'll defer to @conectado to peek over the Rust portion.

**Comment by jamilbk** [comment]
@ReactorScram The pending CI tests that the PR is waiting on happens because the test signatures changed and no longer match our branch protection settings for the repo. It's a non-issue -- I'll update the repo's branch protection rules to match once this PR is ready to merge.

File: .github/workflows/cd.yml
Comment (ReactorScram) [reply .github/workflows/cd.yml:264]: We could add aarch64 Windows in there later on. But the actual reason is that I copy-pasted the Linux build step and wanted this to keep matching the same layout.

Comment (jamilbk) [reply .github/workflows/cd.yml:264]: 👍

**Review by jamilbk — APPROVED** [review]
LGTM, just waiting on review from @conectado

File: rust/windows-client/README.md
Comment (conectado) [inline rust/windows-client/README.md]: Does this work? I used to get `ring` compilation errors when doing this on linux.

File: rust/windows-client/build.rs
Comment (conectado) [inline rust/windows-client/build.rs]: I don't know anything about manifests but should we use `Firezone` instead of `Contoso`?

File: rust/windows-client/src/main.rs
Comment (conectado) [inline rust/windows-client/src/main.rs]: even if it's more verbose I think it's better if we get `unsafe` closer to each `unsafe` function, that way we can add specific unsafe comments and anything that isn't `unsfafe` isn't wrapped on it.

File: rust/windows-client/src/main.rs
Comment (conectado) [inline rust/windows-client/src/main.rs]: Prefer explicit includes to start includes, it's easier to find where symbols come from.

File: rust/windows-client/src/main.rs
Comment (conectado) [inline rust/windows-client/src/main.rs]: Is there any reason for this comment?

File: rust/windows-client/src/main.rs
Comment (conectado) [inline rust/windows-client/src/main.rs]: We could move this to a separate function. We could have one for window creation another for tray, it'll be more readable.

File: rust/windows-client/src/main.rs
Comment (conectado) [inline rust/windows-client/src/main.rs]: I think [this](https://learn.microsoft.com/en-us/windows/win32/api/winuser/ns-winuser-wndclassexa) is the recommended struct now.

File: rust/windows-client/src/main.rs
Comment (conectado) [inline rust/windows-client/src/main.rs]: Seems like this function has been superseded by [this](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-loadimagew)

File: rust/windows-client/src/main.rs
Comment (conectado) [reply rust/windows-client/src/main.rs]: Though, I think we don't use any of the newly provided members.

File: rust/windows-client/src/main.rs
Comment (conectado) [inline rust/windows-client/src/main.rs]: I prefer not using `\0` in an &str since this can lead to some semantic confusion. We can use [`CString`](https://doc.rust-lang.org/std/ffi/struct.CString.html)  instead.

**Review by conectado — COMMENTED** [review]
Awesome work! I left some comments, nothing major.

File: rust/windows-client/build.rs
Comment (ReactorScram) [reply rust/windows-client/build.rs]: Yeah good catch, I think Contoso is a placeholder fake company name.

File: rust/windows-client/src/main.rs
Comment (ReactorScram) [reply rust/windows-client/src/main.rs]: No, just happened to be fighting the LoadIconA function when I committed this one. In my local repo I'm now trying NWG and Tauri so this should be gone next time. I couldn't get LoadIcon to work.

Comment (conectado) [reply rust/windows-client/src/main.rs]: Ah, let's remove it before merging it then.

Also just as a note, I think we tried out tauri before and we decided not to go with it, at least for the other native platforms, I don't exactly remember the reasons maybe @jamilbk remembers.

Comment (ReactorScram) [reply rust/windows-client/src/main.rs]: He said earlier in DM "our reasoning for not going with Tauri was because the it was too much overhead vs just using platform idiomatic approaches on Apple"

I'm trying Tauri now and yeah the RAM usage to do nothing is like 100 MB. But the API is very nice. Up to you two.

Comment (ReactorScram) [reply rust/windows-client/src/main.rs]: @conectado Since I'm just trying to get the CI to produce an exe, could I just cut it down to a "Hello World" CLI? That way the Rust code doesn't slow down this PR. Once the CI/CD is working I can put it back in.

**Review by conectado — APPROVED** [review]
LGTM!

File: rust/windows-client/src/main.rs
Comment (jamilbk) [reply rust/windows-client/src/main.rs]: "too much overhead" was more in reference to engineering overhead (security, troubleshooting, cognitive, maintainability, etc)

Comment (jamilbk) [reply rust/windows-client/src/main.rs]: I suppose what I fear is architecting the app around Tauri then finding out later it doesn't work well for low-level interfacing or something and then having to unwind the app architecture

Comment (jamilbk) [reply rust/windows-client/src/main.rs]: But I guess at the end of the day it's all just Rust and the UI layer is pretty thin, so we should be able to refactor to use other UI libs when we get further along. Probably better to go with whatever's fastest easiest for you @ReactorScram since maintainability isn't a worry for some time

**Comment by jamilbk** [comment]
Updating to resolve conflict

**Pull request merged**
